datalake-core/handlers/Git.cpp

```diff
@@ -0,0 +1,272 @@
+
+    if (!sfs::exists(tgtPath)) {
+      LOG_ERROR << "Storage not initialized, use nmdl-initialize\n";
+      return false;
+    }
+
+    sfs::current_path(tgtPath);
+    return true;
+  }
+
+  bool
+  Git::alignRepo(const nmco::Time& _dts)
+  {
+    LOG_DEBUG << "Target time: " << _dts << '\n';
+
+    std::ostringstream oss;
+    oss << "echo -n `git rev-list -n 1 --first-parent"
+        << " --before=\"" << _dts << "\" master`" ;
+    auto result {nmcu::trim(cmdExecOut(oss.str()))};
+    LOG_DEBUG << "Target SHA: " << result << '\n';
+
+    oss.str("");
+    if (result.empty()) {
+      oss << "git log --reverse --date='format-local:%FT%T'"
+            << " --format=\"format:%cd\""
+          ;
+      auto const& validDates {cmdExecOut(oss.str())};
+      LOG_ERROR << "Invalid repository date: " << _dts << '\n'
+                << "Valid dates:"
+                << '\n' << validDates
+                << '\n';
+      return false;
+    } else {
+      oss << "git checkout -q "
+          << ("infinity" == _dts.toString() ? "master" : result);
+      cmdExec(oss.str());
+    }
+
+    return true;
+  }
+
+  void
+  Git::setIngestToolData(nmdlo::DataEntry& _de, const std::string& _path)
+  {
+    std::ostringstream oss;
+    oss << "git log -n 1 --pretty=format:\"%B\" -- " << _path;
+
+    std::istringstream iss(cmdExecOut(oss.str()));
+
+    std::regex toolRegex('^' + INGEST_TOOL_PREFIX + "(.*)$");
+    std::regex argsRegex('^' + TOOL_ARGS_PREFIX + "(.*)$");
+    std::smatch m;
+
+    for (std::string line; std::getline(iss, line);) {
+      if (std::regex_search(line, m, toolRegex)) {
+        _de.setIngestTool(m.str(1));
+      }
+      if (std::regex_search(line, m, argsRegex)) {
+        _de.setToolArgs(m.str(1));
+      }
+    }
+  }
+
+  void
+  Git::initialize()
+  {
+    const auto& tgtPath {this->dataLakePath};
+
+    LOG_DEBUG << "Removing existing: " << tgtPath << '\n';
+    sfs::remove_all(tgtPath);
+    LOG_DEBUG << "Creating new: " << tgtPath << '\n';
+    sfs::create_directories(tgtPath);
+
+    sfs::current_path(tgtPath);
+    std::ostringstream oss;
+    oss << "git init";
+    cmdExec(oss.str());
+
+  }
+
+  void
+  Git::commit(nmdlo::DataEntry& _de)
+  {
+    if (!changeDirToRepo()) { return; }
+
+    // Ensure device directory exists
+    const sfs::path devicePath {this->dataLakePath/_de.getDeviceId()};
+    sfs::create_directories(devicePath);
+
+    // Copy file to store, properly named
+    const sfs::path dstPath {devicePath/_de.getSaveName()};
+    const std::string dstRelPath {sfs::relative(dstPath).string()};
+    sfs::copy(_de.getDataPath(), dstRelPath, sfs::copy_options::overwrite_existing);
+
+    // Store data
+    std::ostringstream oss;
+    oss << "git add ."
+        << " && git commit -m '"
+          << CHECK_IN_PREFIX << dstRelPath
+          << '\n' << INGEST_TOOL_PREFIX << _de.getIngestTool()
+          << '\n' << TOOL_ARGS_PREFIX << _de.getToolArgs()
+        << '\n';
```

Comment:
Doesn't this still need a single quote at the end to correctly terminate the `git commit -m '...'` command?